rust/ql/lib/codeql/rust/controlflow/internal/Scope.qll

```diff
@@ -5,11 +5,32 @@ private import codeql.rust.elements.internal.generated.ParentChild
 
 /**
  * A control-flow graph (CFG) scope.
- *
- * A CFG scope is a callable with a body.
  */
-class CfgScope extends Callable {
-  CfgScope() {
+abstract private class CfgScopeImpl extends AstNode {
+  abstract predicate scopeFirst(AstNode first);
+
+  abstract predicate scopeLast(AstNode last, Completion c);
+}
+
+final class CfgScope = CfgScopeImpl;
+
+class AsyncBlockScope extends CfgScopeImpl, BlockExpr {
```

Comment:
If you add `instance ExprTrees::AsyncBlockExprTree` after the `extends` clause, then you can remove the charpred, and replace the `this.(ExprTrees::AsyncBlockExprTree)` terms with `super`.